src/java.base/share/classes/java/util/stream/Gatherer.java

```diff
@@ -0,0 +1,571 @@
+         * Pushes, if possible, the provided element to the next stage in the
+         * pipeline.
+         *
+         * <p>If this method returns {@code false} then the next stage does
+         * not accept any more elements.
+         *
+         * @param element the element to send
+         * @return {@code true} if more elements can be sent,
+         *         and {@code false} if not.
+         */
+        boolean push(T element);
+
+        /**
+         * Allows for checking whether the next stage is known to not want
+         * any more elements sent to it.
+         *
+         * @apiNote This is best-effort only, once this returns true it should
+         *          never return false again for the same instance.
+         *
+         * <p>By default this method returns {@code false}.
+         *
+         * @return {@code true} if this Downstream is known not to want any
+         *         more elements sent to it, {@code false} if otherwise
+         */
+        default boolean isRejecting() { return false; }
+    }
+
+    /**
+     * An Integrator receives elements and processes them,
+     * optionally using the supplied state, and optionally sends incremental
+     * results downstream.
+     *
+     * @param <A> the type of state used by this integrator
+     * @param <T> the type of elements this integrator consumes
+     * @param <R> the type of results this integrator can produce
+     * @since 22
+     */
+    @FunctionalInterface
+    @PreviewFeature(feature = PreviewFeature.Feature.STREAM_GATHERERS)
+    interface Integrator<A, T, R> {
+        /** Integrate is the method which given:
+         * the current state, the next element, and a downstream object;
+         * performs the main logic -- potentially inspecting and/or updating
+         * the state, optionally sending any number of elements downstream
+         * -- and then returns whether more elements are to be consumed or not.
+         *
+         * @param state The state to integrate into
+         * @param element The element to integrate
+         * @param downstream The downstream object of this integration,
+         *                   returns false if doesn't want any more elements
+         * @return {@code true} if subsequent integration is desired,
+         *         {@code false} if not
+         */
+        boolean integrate(A state, T element, Downstream<? super R> downstream);
+
+        /**
+         * Factory method for turning Integrator-shaped lambdas into
+         * Integrators.
+         *
+         * @param integrator a lambda to create as Integrator
+         * @return the given lambda as an Integrator
+         * @param <A> the type of elements this integrator receives
+         * @param <T> the type of initializer this integrator consumes
+         * @param <R> the type of results this integrator can produce
+         */
+        @ForceInline
+        static <A, T, R> Integrator<A, T, R> of(Integrator<A, T, R> integrator) {
```

Comment:
While this idiom is very convenient, there are some trap doors in combinations with method references and compatibility which is why we backed out of https://github.com/openjdk/jdk/pull/16213

It is good that the text mentions "lambda" but maybe we should explicitly mention that method references might introduce said problems?

Reply:
This idiom is safer than casts since the casts are blind and need unchecked-annotations. I'm willing to take this approach for an evaluation during the Preview phase.

Reply:
I like the concept but we could document how to safely use it.